### ISSUE TITLE
stability: cluster unable to recover after 2 node outage

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#15591]
Forked from [#15026](https://github.com/cockroachdb/cockroach/issues/15026#issuecomment-298589102):

We have another issue. After running a cluster for 10 days two nodes died during weekend. It's a struggle to get them up again. Lots of errors like below:

My question is: Should I make this another issue? Is this us doing things wrong? How do we debug this? 

We really love Cockroach, but there is  a lot of things that goes wrong for us at the moment. 

```
I170502 10:01:52.180016 76 storage/replica_raftstorage.go:413  [replicate,n3,s3,r29/3:/System/tsd/cr.node.sys.cpu.s…,@c4204ffb00] generated preemptive snapshot 40410405 at index 53
I170502 10:01:52.183406 76 storage/replicate_queue.go:231  [replicate,n3,s3,r29/3:/System/tsd/cr.node.sys.cpu.s…,@c4204ffb00] snapshot failed: r29: remote declined snapshot: reservation rejected
```
AND 

```
I170502 09:59:21.546561 31 server/status/runtime.go:227  [n1] runtime stats: 99 MiB RSS, 209 goroutines, 17 MiB/8.3 MiB/32 MiB GO alloc/idle/total, 13 MiB/19 MiB CGO alloc/total, 68.10cgo/sec, 0.01/0.00 %(u/s)time, 0.00 %gc (0x)
I170502 09:59:22.121205 252 vendor/google.golang.org/grpc/clientconn.go:806  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 10.135.15.39:26257: getsockopt: connection refused"; Reconnecting to {10.135.15.39:26257 <nil>}
E170502 09:59:22.704451 36 gossip/gossip.go:972  [n1] unable to get address for node 3: unable to look up descriptor for node 3
E170502 09:59:22.704851 36 gossip/gossip.go:972  [n1] unable to get address for node 3: unable to look up descriptor for node 3
I170502 09:59:22.955127 252 vendor/google.golang.org/grpc/clientconn.go:806  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 10.135.15.39:26257: getsockopt: connection refused"; Reconnecting to {10.135.15.39:26257 <nil>}
```

**Comment by petermattis** [comment]
@mattiasmak The `reservation rejected` log message can happen in normal operation and isn't a huge concern in and of itself. 

The `unable to get address for node 3` is more worrisome and likely related to the real problem. I'm assuming these log messages are from 2 different machines. The first thing to eliminate is a networking problem. Can node 1 (`n1`) talk to 10.135.15.39? Can you telnet/nc from `n1` to 10.135.15.39:26257? Is that IP address the actual address of `n3`?

**Comment by mattiasmak** [comment]
HI! It seems like I can connect. And it's the actual address. 

<img width="338" alt="screen1" src="https://cloud.githubusercontent.com/assets/606639/25624456/b90a279e-2f59-11e7-8254-86f8f004e392.png">

**Comment by petermattis** [comment]
@mattiasmak Ok. We're going to need more info in order to debug this. Can you share the full log files from the nodes in the cluster?

**Comment by mattiasmak** [comment]
I can share what ever you want. This is an insecure cluster. http://46.101.128.234:8080/#/cluster/all/overview

I can also give ssh access if that makes things easier. You could generate an ssh key that I can add to the nodes.

Here's the logs:

[RO-DB03-logs.tar.gz](https://github.com/cockroachdb/cockroach/files/972613/RO-DB03-logs.tar.gz)
[RO-DB02-logs.tar.gz](https://github.com/cockroachdb/cockroach/files/972614/RO-DB02-logs.tar.gz)
[RO-DB01-logs.tar.gz](https://github.com/cockroachdb/cockroach/files/972612/RO-DB01-logs.tar.gz)

**Comment by mattiasmak** [comment]
After rebooting all nodes again and starting one by one they seems to work... So perhaps the logs is the best clue to what goes wrong.

**Comment by petermattis** [comment]
@mattiasmak Glad to here the cluster is working again. And thanks for the logs. I'll take a look at them later this morning.

**Comment by mattiasmak** [comment]
I can see now that some tables is not working. selecting from a simple table called whitelist is not responding at all. (Waited for 500seconds+)

Also create table does the same thing to me. No response at all.

SELECT * FROM whitelist;

CREATE TABLE whitelist2 (
	prefix STRING PRIMARY KEY,
	os STRING NULL,
	last_seen TIMESTAMPTZ NULL
);

**Comment by mattiasmak** [comment]
I'm serious about this problem. Trying to reboot node 3 to see what happens since I can't to SELECT queries. (see above) The cluster goes down again.

This has happened to us twice and I think this could easily happen to others - We do nothing special. This is not a small issue.

You can access the nodes if you want. Just tell my what to do. Add a ssh key?

**Comment by petermattis** [comment]
@mattiasmak I'm looking at your logs right now. I'll let you know if I need access to the cluster.

**Comment by petermattis** [comment]
@mattiasmak Looks like there are still network issues between the nodes. Getting ssh access to the nodes would be great. Where should I send an ssh key?

**Comment by mattiasmak** [comment]
You've got mail.

**Comment by mattiasmak** [comment]
Current status: 

After a few hours, we can't do selects on the DB. If we reboot all nodes the cluster seems to work again for a few hours. 

Currently it's in a bad state. I will keep it like this until you have time to check.

**Comment by petermattis** [comment]
For posterity: @mattiasmak is running on machines with 2 IP addresses. The nodes are being started with `--advertise-host` to restrict traffic to one of the IP addresses. But somehow the hostnames are being exchanged. The hostnames are not resolvable so the nodes then can't talk to each other. Specifying `--host` works around this bug, but we need to figure out where we're leaking the hostname despite `--advertise-host` being specified.

Cc @a-robinson and @tamird as you are more familiar with this code than I am.

**Comment by tamird** [comment]
Naively, I can't reproduce this. I'm running two nodes in two docker containers, neither of which can resolve the other's hostname.

node 1:
```
$ docker run -p 8080:8080 --rm --volume=$(pwd)/cockroach-linux-2.6.32-gnu-amd64:/coach -it debian bash

$ ./cockroach start -s type=mem,size=1GiB --insecure --logtostderr --advertise-host 172.17.0.4
```

node 2:
```
$ docker run --rm --volume=$(pwd)/cockroach-linux-2.6.32-gnu-amd64:/cockroach -it debian bash

$ ./cockroach start -s type=mem,size=1GiB --insecure --logtostderr --advertise-host 172.17.0.5 --join 172.17.0.4
```

I'm able to look at the first container's admin UI and everything seems to be working. Have I missed something? Is this happening on a particular version?

**Comment by petermattis** [comment]
The other detail is the 2 IP addresses. `eth0` was bound to the external address and `eth1` to the internal address (i.e. the addresses being advertised in `--advertise-host`). Not sure why that would matter.

**Comment by a-robinson** [comment]
@tamird we ran into this while testing a cross-cloud multi-Kubernetes-cluster setup today when testing on top of rc1 when it hadn't been seen in previous tests on older versions (I don't know which specific older versions). We should figure this out ASAP since it appears to be a regression, possibly related to recent flag and gossip changes. Are you still working on this?

**Comment by a-robinson** [comment]
I wasn't able to save much from that cluster since it got taken down before I could, but I did get this PDF of the logs from one of the confused nodes: [advertise-addr-node2.pdf](https://github.com/cockroachdb/cockroach/files/979870/advertise-addr-node2.pdf)

**Comment by a-robinson** [comment]
The issue was that nodes in different cloud providers had the same hostnames but different IPs. The nodes in each cloud with the overlapping hostnames seemed to be getting mixed up with each other, i.e. `cockroachdb-worker-1` in cloud A was getting mixed up with `cockroachdb-worker-1` in cloud B.

**Comment by tamird** [comment]
Ouch. Do you have a pointer to how the setup was being done? In particular, what flags were in use?

I'm not currently working on this actively so if you'd rather take it over, let me know. Otherwise I'm happy to help however I can. Let's discuss in Slack?

**Comment by tamird** [comment]
Followed up with @a-robinson and tried to simulate his scenario using docker - two nodes which can resolve each other's hostnames, with a third that can't resolve either hostname but has the same name as the second. Nodes were started with:
```
roach1: /cockroach start -s type=mem,size=1GiB --insecure --logtostderr --advertise-host 172.18.0.2 --http-host 0.0.0.0

roach2: /cockroach start -s type=mem,size=1GiB --insecure --logtostderr --advertise-host 172.18.0.3 --http-host 0.0.0.0 --join 172.18.0.2

roach2-again: /cockroach start -s type=mem,size=1GiB --insecure --logtostderr --advertise-host 172.18.0.4 --http-host 0.0.0.0 --join 172.18.0.2
```
I also ran this without the `--http-host` bit in an earlier iteration. But I wasn't able to observe any problems.

AFAIK, a node only discovers its hostname when it calls `net.Listen()` and then `.Addr()` on the result, which is only done in server/server.go, and that value doesn't ultimately go anywhere. Only the advertised address makes it to gossip and kv. I'm at a loss. Perhaps SSH access would help. @petermattis @mattiasmak? My public key: `curl -sfSL https://api.github.com/users/tamird/keys | python -c 'import sys, json; print json.load(sys.stdin)[0]["key"]'`.

**Comment by a-robinson** [comment]
I can't say whether it's the same thing that was going on in @mattiasmak's cluster, but I figured out and was able to reproduce what was happening in the cross-cloud Kubernetes case.

* At some point before I got involved, someone had set up a cross-cloud cluster in clouds A and B. After their testing was done, they took down the nodes in A but left up the nodes in B.
* Later, while I was around, we brought up a new cluster in node A *with the exact same IP addresses*.
* Although the nodes in B from the old cluster couldn't properly join the new cluster via gossip due to the cluster ID check, they **could** still talk to the nodes in A because they were at the same IP address. In practice, it looks like the old nodes leftover in B do two main things:
   * Open raft transport streams and send messages to the new nodes in A. If the node/store ID of the node in A doesn't match the expectation of the node in B, all that happens is that requests get rejected. However, if the ID *does* match, then bad things can happen, like the crash in the PDF I attached [above](https://github.com/cockroachdb/cockroach/issues/15591#issuecomment-299530273) in which one of the new nodes crashed because it didn't have a raft group in place for the request.
   * Try to update the node liveness table. This is presumably why the IPs on the node list page of the UI would change every time we refreshed it -- sometimes the new node would have most recently updated the liveness and sometimes the old node would have.

In other words, we need to improve how we handle situations like this for 1.1, but I don't see any recent regression or major bug that user error wouldn't avoid.

**Comment by a-robinson** [comment]
Sorry for polluting this issue -- the thing I ran into might not have been the same. I've split off the cross-cloud IP reuse problem to https://github.com/cockroachdb/cockroach/issues/15801.

**Comment by tamird** [comment]
I'm going to give this a close; without further information or access to the machines in question I'm not sure what else can be done.
